File: tests/test_performance_outcome.py
#!/usr/bin/env python3
"""Synthetic measurement controls, not performance measurements."""
import copy
import csv
import json
import os
from pathlib import Path
import platform
import shutil
import subprocess
import sys
import tempfile
import unittest

ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "examples/planning/performance"



class PerformanceOutcome(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory(prefix="hydra-performance-controls-")
        self.addCleanup(self.temp.cleanup)
        self.base = Path(self.temp.name)
        self.inputs = self.base / "inputs"
        self.outputs = self.base / "outputs"
        self.inputs.mkdir()
        self.outputs.mkdir()
        shutil.copy(SOURCE / "records.txt", self.inputs / "records")
        self.context = self.base / "validation.json"
        self.context.write_text(json.dumps({"data": {"performance-check": "1" * 64,
                                                   "performance-check-recipe": "2" * 64}}))
        self.env = {**os.environ, "HYDRA_WORKFLOW_INPUTS_DIR": str(self.inputs),
                    "HYDRA_WORKFLOW_OUTPUTS_DIR": str(self.outputs), "HYDRA_WORKFLOW_REPO_ROOT": str(SOURCE),
                    "HYDRA_WORKFLOW_VALIDATION_FILE": str(self.context)}
        self.manifest, self.rows = self.fixture(.6)

    def fixture(self, multiplier):
        import hashlib
        digest = lambda path: hashlib.sha256(path.read_bytes()).hexdigest()
        tools = {"shell": shutil.which("sh"), "awk": shutil.which("awk"), "python": sys.executable}
        manifest = {"schema_version": 1, "workload": {"path": "records.txt",
                    "sha256": digest(self.inputs / "records"), "bytes": (self.inputs / "records").stat().st_size},
                    "commands": {name: [tools["shell"], str(SOURCE / f"{name}.sh"), str(self.inputs / "records")]
                                 for name in ("baseline", "candidate")},
                    "sources": {name: {"path": f"{name}.sh", "sha256": digest(SOURCE / f"{name}.sh"),
                                      "bytes": (SOURCE / f"{name}.sh").stat().st_size} for name in ("baseline", "candidate")},
                    "environment": {"python": sys.version, "platform": platform.platform(), "machine": platform.machine(),
                    "cwd": str(SOURCE), "timer": "time.perf_counter_ns", "units": "nanoseconds",
                    "source_git": {name: subprocess.check_output(["git", "-C", str(SOURCE), "rev-parse", ref], text=True).strip()
                                   for name, ref in (("commit", "HEAD"), ("tree", "HEAD^{tree}"))},
                    "tools": {name: {"path": path, "sha256": digest(Path(path))} for name, path in tools.items()},
                    "load_before": [0., 0., 0.], "load_after": [0., 0., 0.]},
                    "protocol": {"warmups": 2, "trials": 10, "order": "alternating_AB_BA",
                    "stopping_rule": "exactly_10_pairs_no_exclusions_or_retries", "sample_timeout_seconds": 10,
                    "expected_count": "4003", "exclusions": [], "exclusive_hydra_measurement": True,
                    "exclusivity_scope": "coordinated Hydra jobs; other system activity is observed, not excluded"},
                    "warmups": [{"implementation": name, "warmup": n, "elapsed_ns": 1000, "status": "ok",
                                 "count": "4003", "returncode": 0} for n in (1, 2) for name in ("baseline", "candidate")],
                    "failures": []}
        rows = []
        for trial in range(1, 11):
            names = ("baseline", "candidate") if trial % 2 else ("candidate", "baseline")
            for position, name in enumerate(names):
                rows.append({"sample_id": f"{trial:02d}-{name}", "implementation": name, "trial": str(trial),
                             "order": str(position), "elapsed_ns": str(int((1000 + trial * 13) * (multiplier if name == "candidate" else 1))),
                             "status": "ok", "count": "4003", "returncode": "0"})
        return manifest, rows

    def compose(self, manifest=None, rows=None):
        manifest = self.manifest if manifest is None else manifest
        rows = self.rows if rows is None else rows
        (self.inputs / "manifest").write_text(json.dumps(manifest))
        with (self.inputs / "raw").open("w", newline="") as handle:
            writer = csv.DictWriter(handle, fieldnames=list(self.rows[0]))
            writer.writeheader()
            writer.writerows(rows)
        subprocess.run([sys.executable, str(SOURCE / "analyze.py")], env=self.env,
                       check=True, capture_output=True, text=True, timeout=20)
        shutil.copy2(self.outputs / "report.json", self.inputs / "subject")
        return json.loads((self.inputs / "subject").read_text())

    def check(self, success):
        result = subprocess.run([sys.executable, str(SOURCE / "checker.py")], cwd=SOURCE,
                                env=self.env, capture_output=True, text=True, timeout=20)
        self.assertEqual(result.returncode, 0 if success else 1, result.stdout + result.stderr)
        report = json.loads((self.outputs / "assessment").read_text())
        self.assertEqual(report["schema_version"], 3)
        self.assertEqual(report["domain_verdict"], "pass" if success else "fail")
        self.assertEqual(len(report["evidence_records"]), 5)
        for record in report["evidence_records"]:
            self.assertEqual(record["counts"], {"executed": 1, "failed": 0 if success else 1, "skipped": 0})
            self.assertIsInstance(record["observations"][0]["raw"]["measurement"], int)
        return report

    def test_positive_and_valid_no_improvement(self):
        report = self.compose()
        self.assertEqual(report["outcome"], "target established")
        self.check(True)
        self.manifest, self.rows = self.fixture(1.)
        report = self.compose()
        self.assertEqual(report["outcome"], "target not established")
        self.assertEqual(report["paired_uncertainty"]["estimate"], 0)
        self.check(True)

    def test_exact_measurement_domains(self):
        for fault in ("truncated", "duplicate", "wrong_count", "nonzero", "zero_time", "bad_order"):
            with self.subTest(fault=fault):
                rows = copy.deepcopy(self.rows)
                if fault == "truncated": rows.pop()
                elif fault == "duplicate": rows[1] = rows[0]
                elif fault == "wrong_count": rows[0]["count"] = "4002"
                elif fault == "nonzero": rows[0]["returncode"] = "7"
                elif fault == "zero_time": rows[0]["elapsed_ns"] = "0"
                else: rows[0]["order"] = "1"
                report = self.compose(rows=rows)
                self.assertEqual(report["outcome"], "invalid/insufficient measurement")
                self.check(False)

    def test_source_protocol_and_warmup_binding(self):
        for fault in ("source", "source_git", "workload", "units", "protocol", "missing_warmup", "failed_warmup", "failed_record"):
            with self.subTest(fault=fault):
                manifest = copy.deepcopy(self.manifest)
                if fault == "source": manifest["sources"]["baseline"]["sha256"] = "0" * 64
                elif fault == "source_git": manifest["environment"]["source_git"]["tree"] = "0" * 40
                elif fault == "workload": manifest["workload"]["bytes"] += 1
                elif fault == "units": manifest["environment"]["units"] = "milliseconds"
                elif fault == "protocol": del manifest["protocol"]["stopping_rule"]
                elif fault == "missing_warmup": manifest["warmups"] = []
                elif fault == "failed_warmup": manifest["warmups"][0]["returncode"] = 7
                else: manifest["failures"] = [{"phase": "warmup", "returncode": 7}]
                self.compose(manifest=manifest)
                self.check(False)

    def test_tampered_subject_and_raw_are_rejected(self):
        for fault in ("manifest", "raw", "claim", "interval", "limits", "unknown_field"):
            with self.subTest(fault=fault):
                report = self.compose()
                if fault == "manifest": report["manifest"]["commands"]["baseline"][0] = "/different/tool"
                elif fault == "raw": report["raw_samples"][1] += "0"
                elif fault == "claim": report["outcome"] = "target not established"
                elif fault == "interval": report["paired_uncertainty"]["upper"] = -100
                elif fault == "limits": report["limits"] = ["General speedup established"]
                else: report["hidden_exclusion"] = 1
                (self.inputs / "subject").write_text(json.dumps(report))
                self.check(False)

    def test_measurement_requires_coordinated_window(self):
        env = dict(self.env)
        env.pop("HYDRA_PERFORMANCE_EXCLUSIVE", None)
        result = subprocess.run([sys.executable, str(SOURCE / "measure.py")], env=env,
                                capture_output=True, text=True, timeout=5)
        self.assertNotEqual(result.returncode, 0)
        self.assertIn("coordinated measurement window", result.stderr)
        self.assertFalse((self.outputs / "raw.csv").exists())


if __name__ == "__main__":
    unittest.main()

File: tests/native/test_performance_outcome.c
#include "performance_fixture.h"
static struct outcome_fixture f;
static char subject[F_PATH], raw[F_PATH];
static json_object *compose(json_object *manifest, json_object *rows) {
  char path[F_PATH];
  nt_path(path, f.inputs, "manifest");
  nt_json_write(path, manifest);
  measurement_csv(raw, rows, false, false);
  outcome_check("performance-analyze", 0);
  nt_path(path, f.outputs, "report.json");
  nt_copy_tree(path, subject);
  json_object *r = f_read_json(subject, 1000000);
  assert(r);
  return r;
}
static void check(bool success) {
  outcome_check("performance-check", success ? 0 : 1);
  char path[F_PATH];
  nt_path(path, f.outputs, "assessment");
  json_object *r = f_read_json(path, 1000000);
  assert(r);
  nt_equal(f_field(r, "schema_version"), "3");
  assert(!strcmp(f_string(r, "domain_verdict"), success ? "pass" : "fail"));
  json_object *records = f_field(r, "evidence_records");
  assert(json_object_array_length(records) == 5);
  for (size_t i = 0; i < 5; i++) {
    json_object *rec = json_object_array_get_idx(records, i);
    nt_equal(f_field(rec, "counts"),
             success ? "{\"executed\":1,\"failed\":0,\"skipped\":0}"
                     : "{\"executed\":1,\"failed\":1,\"skipped\":0}");
    assert(json_object_is_type(
        f_field(
            f_field(json_object_array_get_idx(f_field(rec, "observations"), 0),
                    "raw"),
            "measurement"),
        json_type_int));
  }
  json_object_put(r);
}
static void positive(json_object *m, json_object *rows) {
  json_object *r = compose(m, rows);
  assert(!strcmp(f_string(r, "outcome"), "target established"));
  json_object_put(r);
  check(true);
  json_object *same = measurement_rows(1.0);
  r = compose(m, same);
  assert(!strcmp(f_string(r, "outcome"), "target not established"));
  assert(json_object_get_double(
             f_field(f_field(r, "paired_uncertainty"), "estimate")) == 0);
  json_object_put(r);
  json_object_put(same);
  check(true);
}
static void invalid_rows(json_object *m, json_object *rows) {
  for (size_t i = 0; i < 7; i++) {
    json_object *changed = nt_clone(rows),
                *first = json_object_array_get_idx(changed, 0);
    if (i == 0)
      assert(!json_object_array_del_idx(changed, 19, 1));
    if (i == 1)
      assert(!json_object_array_put_idx(changed, 1, nt_clone(first)));
    if (i == 2)
      f_string_add(first, "count", "4002");
    if (i == 3)
      f_string_add(first, "returncode", "7");
    if (i == 4)
      f_string_add(first, "elapsed_ns", "0");
    if (i == 5)
      f_string_add(first, "order", "1");
    if (i == 6)
      f_string_add(first, "elapsed_ns", "9223372036854775807");
    json_object *r = compose(m, changed);
    assert(!strcmp(f_string(r, "outcome"), "invalid/insufficient measurement"));
    json_object_put(r);
    json_object_put(changed);
    check(false);
  }
}
static void manifest_faults(json_object *m, json_object *rows) {
  for (size_t i = 0; i < 8; i++) {
    json_object *v = nt_clone(m), *env = f_field(v, "environment"),
                *warmups = f_field(v, "warmups");
    if (i == 0)
      f_string_add(
          f_field(f_field(v, "sources"), "baseline"), "sha256",
          "0000000000000000000000000000000000000000000000000000000000000000");
    if (i == 1)
      f_string_add(f_field(env, "source_git"), "tree",
                   "0000000000000000000000000000000000000000");
    if (i == 2) {
      json_object *w = f_field(v, "workload");
      json_object_object_add(
          w, "bytes",
          json_object_new_int64(json_object_get_int64(f_field(w, "bytes")) +
                                1));
    }
    if (i == 3)
      f_string_add(env, "units", "milliseconds");
    if (i == 4)
      json_object_object_del(f_field(v, "protocol"), "stopping_rule");
    if (i == 5)
      json_object_object_add(v, "warmups", json_object_new_array());
    if (i == 6)
      json_object_object_add(json_object_array_get_idx(warmups, 0),
                             "returncode", json_object_new_int(7));
    if (i == 7)
      json_object_object_add(
          v, "failures",
          f_parse_value("[{\"phase\":\"warmup\",\"returncode\":7}]"));
    json_object_put(compose(v, rows));
    json_object_put(v);
    check(false);
  }
}
static void subject_faults(json_object *m, json_object *rows) {
  for (size_t i = 0; i < 6; i++) {
    json_object *r = compose(m, rows);
    if (i == 0)
      assert(!json_object_array_put_idx(
          f_field(f_field(f_field(r, "manifest"), "commands"), "baseline"), 0,
          json_object_new_string("/different/tool")));
    if (i == 1) {
      json_object *a = f_field(r, "raw_samples");
      const char *s = json_object_get_string(json_object_array_get_idx(a, 1));
      size_t n = strlen(s);
      char *changed = malloc(n + 2);
      assert(changed);
      memcpy(changed, s, n);
      changed[n] = '0';
      changed[n + 1] = 0;
      assert(!json_object_array_put_idx(a, 1, json_object_new_string(changed)));
      free(changed);
    }
    if (i == 2)
      f_string_add(r, "outcome", "target not established");
    if (i == 3)
      json_object_object_add(f_field(r, "paired_uncertainty"), "upper",
                             json_object_new_int(-100));
    if (i == 4)
      json_object_object_add(
          r, "limits", f_parse_value("[\"General speedup established\"]"));
    if (i == 5)
      json_object_object_add(r, "hidden_exclusion", json_object_new_int(1));
    nt_json_write(subject, r);
    json_object_put(r);
    check(false);
  }
}
static void csv_boundaries(json_object *m, json_object *rows) {
  json_object_put(compose(m, rows));
  char output[F_PATH];
  nt_path(output, f.outputs, "report.json");
  for (size_t i = 0; i < 2; i++) {
    measurement_csv(raw, rows, i == 0, true);
    outcome_check("performance-analyze", 0);
    nt_copy_tree(output, subject);
    check(i != 0);
  }
  const char *bad[] = {
      "bad,header\n1,2\n",
      "sample_id,implementation,trial,order,elapsed_ns,status,count,"
      "returncode\n\"unterminated,baseline,1,0,1000,ok,4003,0\n"};
  for (size_t i = 0; i < 2; i++) {
    nt_write(raw, bad[i]);
    outcome_check("performance-analyze", 0);
    nt_copy_tree(output, subject);
    json_object *r = f_read_json(subject, 1000000);
    assert(!strcmp(f_string(r, "outcome"), "invalid/insufficient measurement"));
    json_object_put(r);
    check(false);
  }
}
static void measurement_gate(void) {
  assert(!unsetenv("HYDRA_PERFORMANCE_EXCLUSIVE"));
  char *argv[] = {"./plan-example", "performance-measure", NULL};
  struct f_capture c = nt_run(argv);
  assert(c.status && strstr(c.err, "coordinated measurement window"));
  f_capture_free(&c);
  char path[F_PATH];
  nt_path(path, f.outputs, "raw.csv");
  assert(access(path, F_OK));
}
static size_t nul_samples(const char *bytes, size_t size) {
  size_t count = 0;
  for (size_t i = 0; i < size; i++) {
    if (bytes[i] != '\0')
      continue;
    assert(i >= 4 && size - i >= 4);
    assert(!memcmp(bytes + i - 4, "4003\0bad", 8));
    count++;
  }
  return count;
}
static void nul_output(void) {
  /* Deliberately invalid stdout exercises capture, not a speedup experiment. */
  nt_write("baseline.sh", "#!/bin/sh\nprintf '4003\\000bad\\n'\n");
  nt_write("candidate.sh", "#!/bin/sh\nprintf '4003\\n'\n");
  outcome_commit();
  assert(!setenv("HYDRA_PERFORMANCE_EXCLUSIVE", "1", 1));
  outcome_check("performance-measure", 0);
  assert(!unsetenv("HYDRA_PERFORMANCE_EXCLUSIVE"));
  char path[F_PATH], input[F_PATH];
  nt_path(path, f.outputs, "manifest.json");
  json_object *manifest = f_read_json(path, 1000000);
  json_object *failures = f_field(manifest, "failures");
  assert(json_object_array_length(failures) == 12);
  for (size_t i = 0; i < 12; i++) {
    json_object *failure = json_object_array_get_idx(failures, i);
    assert(!strcmp(f_string(failure, "implementation"), "baseline"));
    json_object *count = f_field(failure, "count");
    assert(json_object_get_string_len(count) == 8);
    assert(!memcmp(json_object_get_string(count), "4003\0bad", 8));
  }
  json_object_put(manifest);
  nt_path(input, f.inputs, "manifest");
  nt_copy_tree(path, input);
  nt_path(path, f.outputs, "raw.csv");
  struct stat st;
  assert(!stat(path, &st));
  char *bytes = f_read(path, 1000000);
  assert(bytes && nul_samples(bytes, (size_t)st.st_size) == 10);
  free(bytes);
  nt_copy_tree(path, raw);
  outcome_check("performance-analyze", 0);
  nt_path(path, f.outputs, "report.json");
  nt_copy_tree(path, subject);
  json_object *report = f_read_json(subject, 1000000);
  assert(
      !strcmp(f_string(report, "outcome"), "invalid/insufficient measurement"));
  json_object *lines = f_field(report, "raw_samples");
  assert(json_object_array_length(lines) == 21);
  size_t preserved = 0;
  for (size_t i = 0; i < 21; i++) {
    json_object *line = json_object_array_get_idx(lines, i);
    preserved += nul_samples(json_object_get_string(line),
                             (size_t)json_object_get_string_len(line));
  }
  assert(preserved == 10);
  json_object_put(report);
  check(false);
}
int main(void) {
  outcome_setup(&f, "performance");
  assert(!setenv("HYDRA_WORKFLOW_REPO_ROOT", f.repo, 1));
  char path[F_PATH];
  nt_path(path, f.inputs, "records");
  nt_copy_tree("records.txt", path);
  nt_path(subject, f.inputs, "subject");
  nt_path(raw, f.inputs, "raw");
  nt_path(path, f.folder, "validation.json");
  nt_write(
      path,
      "{\"data\":{\"performance-check\":"
      "\"1111111111111111111111111111111111111111111111111111111111111111\","
      "\"performance-check-recipe\":"
      "\"2222222222222222222222222222222222222222222222222222222222222222\"}}");
  json_object *m = measurement_manifest(&f), *rows = measurement_rows(.6);
  positive(m, rows);
  invalid_rows(m, rows);
  manifest_faults(m, rows);
  subject_faults(m, rows);
  csv_boundaries(m, rows);
  measurement_gate();
  nul_output();
  json_object_put(m);
  json_object_put(rows);
  outcome_finish(&f,
                 "Performance: five original groups plus CSV, integer-range "
                 "and NUL-output regressions passed");
  return 0;
}
